plugin/spi.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"context"
+	"unsafe"
+
+	"github.com/pingcap/tidb/sessionctx/variable"
+)
+
+const (
+	// LibrarySuffix defines TiDB plugin's file suffix.
+	LibrarySuffix = ".so"
+	// ManifestSymbol defines TiDB plugin's entrance symbol.
+	// Plugin take manifest info from this symbol.
+	ManifestSymbol = "PluginManifest"
+)
+
+// Manifest describes plugin info and how it can do by plugin itself.
+type Manifest struct {
+	Kind           Kind
+	Name           string
+	Description    string
+	Version        uint16
+	RequireVersion map[string]uint16
+	License        string
+	BuildTime      string
+	SysVars        map[string]*variable.SysVar
+	Validate       func(ctx context.Context, manifest *Manifest) error
+	OnInit         func(ctx context.Context, manifest *Manifest) error
+	OnShutdown     func(ctx context.Context, manifest *Manifest) error
+}
+
+// ExportManifest exports a manifest to TiDB as a known format.
+// it just casts sub-manifest to manifest.
+func ExportManifest(m interface{}) *Manifest {
+	return (*Manifest)((*(*iface)(unsafe.Pointer(&m))).Data)
```

Comment:
If Manifest is not the first field in the extended manifest struct, this code broke.

Reply:
Yes, but only tidb developer can add new kind of manifest, this can be restricted by code review(and it also has be documented); for plugin develop only touch toml configuration with predefine manifest